### ISSUE TITLE
Using CheckSession in MVC + Angular SPA workflow Question

### ISSUE THREAD

**Issue by phillsta** [IdentityServer/IdentityServer3#2279]
Hi,

Need some help with the checksession feature please (been at this for days)!

I have an angular app sitting inside a mvc project. I want to be able to check the session regularly and display a message require login if the session has "changed". I have no idea where to get the session_state variable for the message that is sent to the OP in postMessage. Help!

Another question if I have the current flow do you see a security risk or can you suggest a better way of doing it
1. Open Application
2. Redirects to IdentityServer3 Login
3. Login Successful, Redirects back to app
4. App builds claims into user session using access_token, expires_at, refresh_token and id_token
5. User loads the SPA which reads the access_token claim and makes it available as a JS object for use in the angular app (for resource api calls)
6. When the calls to the resource API return 401 Unauthorised, app intercepts calls MVC method to refresh the token and get another access_token and recalls resource API.
... and the world is happy

**Comment by brockallen** [comment]
session_state is part of the authorization response. You'd handle this in your OIDC middleware callbacks once the user has been fully validated/authenticated. 

Having said that, this is a very JS-focused approach. You can also just implement an end session callback URL in your app as well. We added this in a recent release.

**Comment by phillsta** [comment]
Thanks mate

That worked but now I got another problem maybe you can help with.

The angularJS uses mvc views but as these pages require security. when the view is accessed [in angular] and the session has expired for example [in using the authorize attribute on the view controller] the redirects and loads the error page inside the identity server.
When and error saying the client is unknown...

Any advice?

Also the page references local script.min.js file which doesn't find when the page is loaded inside an angular view.

Hope this makes sense.

Phill




On Fri, Dec 11, 2015 at 6:05 AM -0800, "Brock Allen" <notifications@github.com<mailto:notifications@github.com>> wrote:


session_state is part of the authorization response. You'd handle this in your OIDC middleware callbacks once the user has been fully validated/authenticated.

Having said that, this is a very JS-focused approach. You can also just implement an end session callback URL in your app as well. We added this in a recent release.

-

**Comment by phillsta** [comment]
I was thinking to setup the mvc as a resource server and get angular to handle authentication that way when it requests a token it can be for the mvc and the api. 

In this case do u have any samples on mvc using a bearer token please?

Phill

**Comment by brockallen** [comment]
If the MVC code requires the user to be authenticated to render the NG views, then make the MVC app a client of IdSvr and get the user authenticated and then issue your cookie from the katana OIDC middleware. 

Also, for new/different questions please open a new issue.

**Comment by phillsta** [comment]
thank, i did try that but im getting this error on the idsvr end .
CORS request made for path: "/login" from origin: "http://localhost:5702" but rejected because invalid CORS path

i've tried all 3 of these configurations with no avail. Any advice?

`
factory.CorsPolicyService = new ClientConfigurationCorsPolicyRegistration(efConfig);
`
`
             factory.CorsPolicyService = new Registration<ICorsPolicyService>(new DefaultCorsPolicyService { AllowAll = true });
`
`
            var cors = new InMemoryCorsPolicyService(Clients.Get());
            factory.CorsPolicyService = new Registration<ICorsPolicyService>(cors);
`

**Comment by phillsta** [comment]
Just an updates.

I ended up incorportating the oidc client and token manager into an angular service that produced the results I was looking for, and I treated the mvc app as a resource [giving it it's own resource scope].

As the views sit within a ng view and are partial renders, the owin cookie method didn't work on redirect to the client as only partial redirect happened resulting in the page on the idsvr erroring out every time.

Thank you for your assistance, your response times and feedback are excellent as always!
